lib/pacemaker/pcmk_sched_actions.c

```diff
@@ -1654,42 +1655,47 @@ process_node_history(pe_node_t *node, xmlNode *lrm_rscs, pe_working_set_t *data_
     }
 }
 
+/*!
+ * \internal
+ * \brief Process any resource configuration changes in the CIB status
+ *
+ * Go through all nodes' resource history, and if a resource's configuration
+ * changed since its actions were done, schedule any actions needed (restart,
+ * reload, unfencing, rescheduling recurring actions, clean-up, etc.).
+ * (This also cancels recurring actions for maintenance mode, which is not
+ * entirely related but convenient to do here.)
+ *
+ * \param[in] data_set  Cluster working set
+ */
 static void
-check_actions(pe_working_set_t * data_set)
+check_actions(pe_working_set_t *data_set)
 {
-    const char *id = NULL;
-    pe_node_t *node = NULL;
-    xmlNode *lrm_rscs = NULL;
     xmlNode *status = pcmk_find_cib_element(data_set->input,
                                             XML_CIB_TAG_STATUS);
-    xmlNode *node_state = NULL;
-
-    for (node_state = pcmk__xe_first_child(status); node_state != NULL;
-         node_state = pcmk__xe_next(node_state)) {
 
-        if (pcmk__str_eq((const char *)node_state->name, XML_CIB_TAG_STATE,
-                         pcmk__str_none)) {
-            id = crm_element_value(node_state, XML_ATTR_ID);
-            lrm_rscs = find_xml_node(node_state, XML_CIB_TAG_LRM, FALSE);
-            lrm_rscs = find_xml_node(lrm_rscs, XML_LRM_TAG_RESOURCES, FALSE);
+    for (xmlNode *node_state = first_named_child(status, XML_CIB_TAG_STATE);
+         node_state != NULL; node_state = crm_next_same_xml(node_state)) {
 
-            node = pe_find_node_id(data_set->nodes, id);
+        pe_node_t *node = NULL;
+        xmlNode *lrm_rscs = NULL;
 
-            if (node == NULL) {
-                continue;
+        lrm_rscs = find_xml_node(node_state, XML_CIB_TAG_LRM, FALSE);
+        lrm_rscs = find_xml_node(lrm_rscs, XML_LRM_TAG_RESOURCES, FALSE);
```

Comment:
It's out of scope for this PR, but these two lines could be an xpath expression that we feed to some other function.

Reply:
Good call. I figured out what I think is a better way of organizing this altogether